### PR TITLE
Adding tabs

### PR DESCRIPTION
**Pull request by dstreett**

### CODE REVIEW

**Comment by dstreett** [comment]
Tabs and Interleaved now. Someone mind looking over it before it goes into master?

File: Super-Deduper/src/super-deduper.cpp [hunk 1]
```diff
@@ -59,6 +59,35 @@ void load_map(InputReader<T, Impl> &reader, Counter& counters, BitMap& read_map,
 }
 
 template <class T>
+void output_read_map_tab(const BitMap& read_map, T& out1) {
+    OutputWriter<ReadBase, ReadBaseOutTab> tabs(out1);
+    for(auto const &i : read_map) {
+        ReadBase* rb = dynamic_cast<ReadBase*>(i.second.get());
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: no need to dynamic cast here, bitmap has readbases in it.

File: Super-Deduper/src/super-deduper.cpp [hunk 2]
```diff
@@ -59,6 +59,35 @@ void load_map(InputReader<T, Impl> &reader, Counter& counters, BitMap& read_map,
...
+        tabs.write(*rb);
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: no need to create the temp varaible, just tabs.write(*(i.second));

File: Super-Deduper/src/super-deduper.cpp [hunk 3]
```diff
@@ -186,6 +217,24 @@ int main(int argc, char** argv)
                     load_map(ifs, counters, read_map, start, length);
                 }
             }
+            
+            if(vm.count("tab-input")) {
+                auto read_files = vm["tab-input"].as<std::vector<std::string> > ();
+                for (auto file : read_files) {
+                    std::ifstream tabReads(file, std::ifstream::in);
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: we'll want to add some file exists checks like the following before opening the ifstream:
    bf::path p(filename);
    if (!bf::exists(p)) {
        throw std::runtime_error("File " + filename + " was not found.");
    }

File: Super-Deduper/src/super-deduper.cpp [hunk 4]
```diff
@@ -186,6 +217,24 @@ int main(int argc, char** argv)
...
+                    InputReader<ReadBase, TabReadImpl> ift(tabReads);
+                    load_map(ift, counters, read_map, start, length);
+                }
+            }
+            
+            if (vm.count("interleaved-input")) {
+                auto read_files = vm["interleaved-input"].as<std::vector<std::string > >();
+                for (auto file : read_files) {
+                    std::ifstream interReads(file, std::ifstream::in);
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: same here, maybe write a function to do that

File: common/src/ioHandler.cpp [hunk 5]
```diff
@@ -39,6 +39,46 @@ Read InputFastq::load_read(std::istream *input) {
     return Read(seq, qual, id.substr(1));
 }
 
+//Overrides load_read for tab delimited reads
+std::vector<Read> TabReadImpl::load_read(std::istream *input) {
+    //to read the line
+    std::string tabLine;
```
Comment (joe-angell) [inline common/src/ioHandler.cpp]: recommend moving this into the TabReadImpl class, that way the buffer will get reuse over each call to load_read

File: common/src/ioHandler.cpp [hunk 6]
```diff
@@ -39,6 +39,46 @@ Read InputFastq::load_read(std::istream *input) {
...
+        
+        if (parsedRead[3].size() != parsedRead[4].size()) {
+            throw std::runtime_error("sequence and qualities are not the same length 2");
+        }
+   
+        reads.push_back(Read(parsedRead[3], parsedRead[4], parsedRead[0]));
+    }
+
+    // ignore extra lines at end of file
+    while(input->good() and input->peek() == '\n') {
```
Comment (joe-angell) [inline common/src/ioHandler.cpp]: watch out for '\r', you get those in windows files, not sure if that is common in this scenario

**Comment by joe-angell** [comment]
Yeah looks good dude, just a couple minor comments.

File: Super-Deduper/src/super-deduper.cpp (hunk 3 above)
Comment (dstreett) [inline Super-Deduper/src/super-deduper.cpp]: Can I just reuse your check_open_r? Or would it be worth it to have a standalone check?

Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: yeah thats fine, that will also decode gz files

File: Super-Deduper/src/super-deduper.cpp [hunk 7]
```diff
@@ -62,8 +62,8 @@ template <class T>
 void output_read_map_tab(const BitMap& read_map, T& out1) {
     OutputWriter<ReadBase, ReadBaseOutTab> tabs(out1);
     for(auto const &i : read_map) {
-        ReadBase* rb = dynamic_cast<ReadBase*>(i.second.get());
-        tabs.write(*rb);
+        //ReadBase* rb = dynamic_cast<ReadBase*>(i.second.get());
```
Comment (joe-angell) [inline Super-Deduper/src/super-deduper.cpp]: remove comment

**Comment by joe-angell** [comment]
looks good other than comment

**Pull request merged**